crux/src/Crux/Goal.hs

```diff
@@ -237,33 +239,66 @@ proveGoalsOffline adapter opts ctx explainFailure (Just gs0) = do
           -- NOTE: We don't currently provide a method for capturing the output
           -- sent to offline solvers.  We would probably want a file per goal.
           let logData = WS.defaultLogData
-          mres <- withTimeout $ WS.solver_adapter_check_sat adapter sym logData [assumptions, goal] $ \satRes ->
-            case satRes of
-              Unsat _ -> do
-                -- NOTE: We don't have an easy way to get an unsat core here
-                -- because we don't have a solver connection.
-                let core = fmap Left (F.toList (mconcat assumptionsInScope))
-                end
-                modifyIORef' goalNum (updateProcessedGoals p (Proved core))
-                return (Prove (p, Proved core))
-              Sat (evalFn, _) -> do
-                let model = ctx ^. cruciblePersonality . personalityModel
-                vals <- evalModel evalFn model
-                explain <- explainFailure evalFn p
-                end
-                let gt = NotProved explain (Just (ModelView vals))
-                modifyIORef' goalNum (updateProcessedGoals p gt)
-                when (failfast && not (isResourceExhausted p)) $
-                  sayOK "Crux" "Counterexample found, skipping remaining goals"
-                return (Prove (p, gt))
-              Unknown -> do
-                end
-                modifyIORef' goalNum (updateProcessedGoals p (NotProved mempty Nothing))
-                return (Prove (p, NotProved mempty Nothing))
-          case mres of
-            Just res -> return res
-            Nothing -> return (Prove (p, NotProved mempty Nothing))
 
+          let runOneSolver adapter = do
```

Comment:
I'm not a big fan of complicated nestings of `where` and `let`.  If we can do it without too much difficulty, I'd like to see this lifted to the same `where` block as `go` (which could really use a better name).